File: cpp/templates/segment_tree.cpp
// t-tree, tl-tree left, tr-tree right, tm-tree mid, v-node, x-new value, p-position
// queries [l,r]
int const MAXN=1e5;
int n,t[4*MAXN];// but requires only 2n-1 space

void build(vector<int>& a,int v,int tl,int tr){
    if(tl==tr)t[v]=a[tl];
    else{
        int tm=tl+(tr-tl)/2;
        build(a,2*v,tl,tm);build(a,2*v+1,tm+1,tr);
        t[v]=t[2*v]+t[2*v+1];
    }
}

int sum(int v,int tl,int tr,int l,int r){
    if(l>r)return 0;
    if(tl==l&&tr==r)return t[v];
    int tm=tl+(tr-tl)/2;
    return sum(v*2,tl,tm,l,min(r,tm))+sum(2*v+1,tm+1,tr,max(l,tm+1),r);
}

void update(int v,int tl,int tr,int p,int x){
    if(tl==tr)t[v]=x;
    else{
        int tm=tl+(tr-tl)/2;
        if(tm<=pos)update(v*2,tl,tm,p,x);
        else update(v*2+1,tm+1,tr,p,x);
        t[v]=t[2*v]+t[2*v+1];
    }
}

int main(){
    n=a.size();
    build(a,1,0,n-1);
    return 0;
}

File: python/segment-tree.py
import os
import io
import sys
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


class SegmentTree:
    def __init__(self, arr):
        n = len(arr)
        self.tree = [0] * (4 * n)
        self.build(arr, 0, 0, n - 1)

    def merge(self, a, b):
        return a + b

    def build(self, arr, node, low, high):
        mid = ((high - low) >> 1) + low
        if low == high:
            self.tree[node] = arr[mid]
        else:
            leftNode, rightNode = (node << 1) + 1, (node << 1) + 2
            self.build(arr, leftNode, low, mid)
            self.build(arr, rightNode, mid + 1, high)
            self.tree[node] = self.merge(
                self.tree[leftNode], self.tree[rightNode])

    def query(self, node, low, high, left, right):
        if high < left or right < low:
            return 0
        if (left <= low and high <= right):
            return self.tree[node]
        mid = ((high - low) >> 1) + low
        leftNode, rightNode = (node << 1) + 1, (node << 1) + 2
        leftSubtree = self.query(leftNode, low, mid, left, right)
        rightSubtree = self.query(rightNode, mid + 1, high, left, right)
        return self.merge(leftSubtree, rightSubtree)

    def update(self, node, low, high, pos, value):
        mid = ((high - low) >> 1) + low
        leftNode, rightNode = (node << 1) + 1, (node << 1) + 2
        if low == high and mid == pos:
            self.tree[node] = value
        elif pos <= mid:
            self.update(leftNode, low, mid, pos, value)
        else:
            self.update(rightNode, mid + 1, high, pos, value)
        self.tree[node] = self.merge(
            self.tree[leftNode], self.tree[rightNode])


def solve():
    n, q = map(int, input().split())
    arr = list(map(int, input().split()))
    # print(n, q, arr)

    segTree = SegmentTree(arr)

    output = []
    for _ in range(q):
        ty, x, y = map(int, input().split())
        if ty == 1:
            pos, value = x, y
            segTree.update(0, 0, n - 1, pos, value)
        else:
            left, right = x, y
            res = segTree.query(0, 0, n - 1, left - 1, right - 1)
            output.append(str(res))
        # sys.stdout.write(str(res) + '\n')

    sys.stdout.write("\n".join(map(str, output)) + "\n")


solve()
